// aleon_35_pdf_ui_lib/stories/Table/Table.stories.jsx
import { ThemeProvider } from '../../ThemeProvider';
import { InformationIcon } from '../Icons/InformationIcon';
import { Tooltip } from '../Tooltip/Tooltip';
import { Table } from './Table';

// More on how to set up stories at: https://storybook.js.org/docs/preact/writing-stories/introduction
export default {
  title: 'Example/Table',
  component: Table,
  tags: ['autodocs'],
  argTypes: {
    leftColumnWidth: { control: 'text' },
    rightColumnWidth: { control: 'text' },
    rows: { control: 'object' },
  },
};

const rowsExample = [
  { left: 'Name', right: <div>Hey<div>Eg</div><br />dr</div> },
  { left: 'Age', right: '25' },
  { left: 'Email', right: 'john.doe@example.com' },
];

// More on writing stories with args: https://storybook.js.org/docs/preact/writing-stories/args
export const Primary = {
  args: {
    leftColumnWidth: "30%",
    rightColumnWidth: "70%",
    rows: rowsExample,
  },
};

export const EvenColumns = {
  args: {
    rows: rowsExample,
  },
};

export const ShortLeft = {
  args: {
    leftColumnWidth: "50px",
    rightColumnWidth: "100%",
    rows: rowsExample,
  },
};

const rowsWithTooltip = [
  { left: 'Name', right: 'Jane Doe' },
  { left: 'Age', right: <InformationIcon /> },
  { left: 'Email', right: <Tooltip title="Click to copy email">john.doe@example.com</Tooltip> },
  { left: 'Age', right: '25' },
];

export const WithTooltip = (args) => (
  <ThemeProvider defaultTheme='dark'>
    <Table leftColumnWidth="30%" rightColumnWidth="70%" rows={rowsWithTooltip} {...args} />
  </ThemeProvider>
);

WithTooltip.args = {
  leftColumnWidth: "30%",
  rightColumnWidth: "70%",
};
