// alien35_pdf_ui_lib_2/stories/Accordion/Accordion.stories.jsx
import { Accordion } from '../../index'; // Adjust path as necessary

// More on how to set up stories at: https://storybook.js.org/docs/preact/writing-stories/introduction
export default {
  title: 'Example/Accordion',
  component: Accordion,
  tags: ['autodocs'],
  argTypes: {
    title: { control: 'text' },
    // Any other props you want to control can go here
  },
};

// More on writing stories with args: https://storybook.js.org/docs/preact/writing-stories/args
const Template = (args) => <Accordion {...args} content={<div>Whaddupp</div>} />;

export const Primary = Template.bind({});
Primary.args = {
  title: "Accordion Title",
  content: <div>Whaddupp</div>
};

export const WithoutTitle = {
  args: {
    title: "",
    content: "This accordion doesn't have a title."
  },
};

export const WithLongContent = {
  args: {
    title: "Accordion with Long Content",
    content: "This is a longer content that might span multiple lines or include additional components and styling."
  },
};
