// alien35_pdf_ui_lib_2/stories/NumberInput/NumberInput.stories.jsx
import { NumberInput, ThemeProvider } from '../../index';

// More on how to set up stories at: https://storybook.js.org/docs/preact/writing-stories/introduction
export default {
  title: 'Example/NumberInput',
  component: NumberInput,
  tags: ['autodocs'],
  argTypes: {
    placeholder: { control: 'text' },
    value: { control: 'number' },
    min: { control: 'number' },
    max: { control: 'number' },
    step: { control: 'number' },
    onChange: { action: 'onChange' },
    backgroundColor: { control: 'color' },
  },
};

// More on writing stories with args: https://storybook.js.org/docs/preact/writing-stories/args
export const PrimaryLight = (args) => (
  <ThemeProvider defaultTheme="light">
    <NumberInput {...args} />
  </ThemeProvider>
);

PrimaryLight.args = {
  placeholder: "0",
  value: 0,
  min: 0,
  max: 9999999,
  step: 1,
};

export const PrimaryDark = (args) => (
  <ThemeProvider defaultTheme="dark">
    <NumberInput {...args} />
  </ThemeProvider>
);

PrimaryDark.args = {
  placeholder: "0",
  value: 0,
  min: 0,
  max: 9999999,
  step: 1,
};

/*

export const Short = {
  args: {
    placeholder: "0",
    value: 0,
    min: 0,
    max: 9999999,
    step: 1,
    width: "10px"
  },
};

export const Long = {
  args: {
    placeholder: "0",
    value: 0,
    min: 0,
    max: 9999999,
    step: 1,
    width: "100px"
  },
};

export const Small = {
  args: {
    placeholder: "0",
    value: 0,
    min: 0,
    size: "sm",
    max: 9999999,
    step: 1,
    width: "100px"
  },
};
*/
