// aleon_35_pdf_ui_lib/stories/Select/Select.stories.jsx
import { Select, ThemeProvider } from '../../index';
const orientationOptions = [
  "Auto",
  { label: "-----------------", options: ["Portrait", "Landscape"] },
];

// More on how to set up stories at: https://storybook.js.org/docs/preact/writing-stories/introduction
export default {
  title: 'Example/Select',
  component: Select,
  tags: ['autodocs'],
  argTypes: {
    options: [],
    backgroundColor: { control: 'color' },
    onClick: { action: 'onClick' },
  },
};

// More on writing stories with args: https://storybook.js.org/docs/preact/writing-stories/args
export const Primary = {
  args: {
    primary: true,
    options: orientationOptions,
    label: 'Add background graphics',
  },
};

export const Short = {
  args: {
    primary: true,
    options: orientationOptions,
    width: "10px",
    label: 'Add background graphics',
  },
};

export const Long = {
  args: {
    primary: true,
    options: orientationOptions,
    width: "200px",
    label: 'Add background graphics',
  },
};

export const Small = {
  args: {
    primary: true,
    options: orientationOptions,
    size: "sm",
    width: "200px",
    label: 'Add background graphics',
  },
};

export const PrimaryDark = (args) => (
  <ThemeProvider defaultTheme="dark">
    <Select {...args} />
  </ThemeProvider>
);

PrimaryDark.args = {
  primary: true,
  options: orientationOptions,
  size: "sm",
  width: "200px",
  label: 'Add background graphics',
};
