// aleon_35_pdf_ui_lib/stories/Button/Button.stories.jsx
import { Button } from './Button';

// More on how to set up stories at: https://storybook.js.org/docs/preact/writing-stories/introduction
export default {
  title: 'Example/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    size: { control: { type: 'select', options: ['sm', 'md', 'lg'] } },
    onClick: { action: 'onClick' },
    label: { control: 'text' },
    backgroundColor: { control: 'color' },
  },
};

// More on writing stories with args: https://storybook.js.org/docs/preact/writing-stories/args

export const LargeSecondary = {
  args: {
    size: 'lg',
    variant: 'secondary',
    label: 'Convert HTML to PDF',
    children: "Hello"
  },
};

export const LargePrimary = {
  args: {
    size: 'lg',
    label: 'Convert HTML to PDF',
    children: "Hello"
  },
};

export const MediumSecondary = {
  args: {
    size: 'md',
    variant: 'secondary',
    label: 'Convert HTML to PDF',
    children: "Hello"
  },
};

export const MediumPrimary = {
  args: {
    size: 'md',
    label: 'Convert HTML to PDF',
    children: "Hello"
  },
};

export const SmallSecondary = {
  args: {
    size: 'sm',
    variant: 'secondary',
    label: 'Convert HTML to PDF',
    children: "Hello"
  },
};

export const SmallPrimary = {
  args: {
    size: 'sm',
    label: 'Convert HTML to PDF',
    children: "Save"
  },
};

export const SmallPrimaryDisabled = {
  args: {
    size: 'sm',
    label: 'Convert HTML to PDF',
    children: "Save",
    disabled: true
  },
};
