// aleon_35_pdf_ui_lib/stories/Tooltip/Tooltip.stories.jsx
import { ThemeProvider } from '../../ThemeProvider';
import { Tooltip } from './Tooltip';

// More on how to set up stories at: https://storybook.js.org/docs/preact/writing-stories/introduction
export default {
  title: 'Example/Tooltip',
  component: Tooltip,
  tags: ['autodocs'],
  argTypes: {
    size: { control: { type: 'select', options: ['sm', 'md', 'lg'] } },
    onClick: { action: 'onClick' },
    label: { control: 'text' },
    backgroundColor: { control: 'color' },
  },
};

// More on writing stories with args: https://storybook.js.org/docs/preact/writing-stories/args

export const Light = (args) => (
  <ThemeProvider>
    <div><Tooltip {...args}>Thank you</Tooltip></div>
    hello
  </ThemeProvider>
);

export const Dark = (args) => (
  <ThemeProvider defaultTheme="dark">
    <Tooltip {...args}>hello</Tooltip>
  </ThemeProvider>
);

Light.args = {
  title: "Hello"
};

Dark.args = {
  title: "Hello"
};

export const Offset = (args) => (
  <ThemeProvider defaultTheme="dark">
    <Tooltip {...args} offsetX='40px'>hello2</Tooltip>
  </ThemeProvider>
);

Offset.args = {
  title: "Hello"
};

export const OffsetLeft = (args) => (
  <ThemeProvider defaultTheme="dark">
    <Tooltip {...args} offsetX='-40px'>hello2</Tooltip>
  </ThemeProvider>
);

OffsetLeft.args = {
  title: "Hello"
};
