// aleon_35_pdf_viewer_lib/webpack.config.js
const path = require('path');
const CopyWebpackPlugin = require('copy-webpack-plugin');

module.exports = {
  entry: './src/index.js',
  output: {
    path: path.resolve(__dirname, 'dist'),
    filename: 'index.js',
    library: 'PdfViewer',
    libraryTarget: 'umd',
    umdNamedDefine: true,
    globalObject: 'this'
  },
  experiments: {
    asyncWebAssembly: true,  // Enable async WebAssembly
    syncWebAssembly: true     // Enable sync WebAssembly
  },
  module: {
    rules: [
      {
        test: /\.css$/,
        use: ['style-loader', 'css-loader'],
      },
      {
        test: /\.(js)$/,
        exclude: /node_modules/,
        use: 'babel-loader',
      },
      {
        test: /\.svg$/,
        use: ['file-loader']
      },
    ]
  },
  resolve: {
    alias: {
      react: "preact/compat",
      'react-dom/test-utils': 'preact/test-utils',
      'react-dom': 'preact/compat',
    },
    extensions: ['.js']
  },
  plugins: [
    new CopyWebpackPlugin({
      patterns: [
        { from: 'lib/pdf.worker.js', to: 'lib/pdf.worker.js' },
        { from: 'lib/pdf_wasm_project.js', to: 'lib/pdf_wasm_project.js' },
        { from: 'lib/pdf_wasm_project_bg.wasm', to: 'lib/pdf_wasm_project_bg.wasm' }
      ]
    })
  ],
  devServer: {
    static: {
      directory: path.join(__dirname, 'dist'),
    },
    compress: true,
    port: 8080,
  },
};
